h2o-core/src/main/java/hex/grid/HyperSpaceSearchCriteria.java

```diff
@@ -89,36 +104,38 @@ public static double default_stopping_tolerance_for_frame(Frame frame) {
     }
 
     public void set_default_stopping_tolerance_for_frame(Frame frame) {
-      _stopping_tolerance = default_stopping_tolerance_for_frame(frame);
-    }
-
-
-    public RandomDiscreteValueSearchCriteria() {
-      super(Strategy.RandomDiscrete);
+      _stopping_criteria._stopping_tolerance = default_stopping_tolerance_for_frame(frame);
     }
 
     public void set_seed(long seed) {
       this._seed = seed;
     }
 
     public void set_max_models(int max_models) {
-      _max_models = max_models;
+      _stopping_criteria._max_models = max_models;
     }
 
     public void set_max_runtime_secs(double max_runtime_secs) {
-      _max_runtime_secs = max_runtime_secs;
+      _stopping_criteria._max_runtime_secs = max_runtime_secs;
     }
 
     public void set_stopping_rounds(int stopping_rounds) {
-      _stopping_rounds = stopping_rounds;
+      _stopping_criteria._stopping_rounds = stopping_rounds;
     }
 
     public void set_stopping_metric(ScoreKeeper.StoppingMetric stopping_metric) {
-      _stopping_metric = stopping_metric;
+      _stopping_criteria._stopping_metric = stopping_metric;
     }
 
     public void set_stopping_tolerance(double stopping_tolerance) {
-      this._stopping_tolerance = stopping_tolerance;
+      _stopping_criteria._stopping_tolerance = stopping_tolerance;
+    }
+
+    @Override
+    public RandomDiscreteValueSearchCriteria deepClone() {
```

Comment:
I dont like this method, i think since we have separate SearchCriteria class it should be immutable, in which case cloning is not necessary, rather than having copy-modifier method like withStoppingTolerance(value) 
also since we are refactoring this could we change the setters to camelCase?

Reply:
@honzasterba I'm not a big fan either to be honest (see https://github.com/h2oai/h2o-3/pull/4313#discussion_r380367954), but the only alternative is "constructor copy", e.g.
```
SearchCriteria(SearchCriteria original) {
   //loop through fields
}
```
AutoML (and a custom grid could do that as well), adaptively changes some of the search criteria, and for this we need to copy the original and change the new ones (for example `max_runtime_secs`) so immutability is not possible here... or it is if we simply create a constructor taking all the criteria... I'm open to this.
Note though that `H2O-3` makes heavy use of reflection, and therefore immutability is never 100% certain.

> also since we are refactoring this could we change the setters to camelCase?
Not here, this is part of public API.

Reply:
a copy-constructor came to my mind as well, even that I like better than deepClone

Reply:
OK, some can be converted into camelCase indeed.

Reply:
I may have a solution to get rid of the deepClone as well, and make `StoppingCriteria` not completely immutable, but at least partially.

Reply:
@honzasterba removed, now cloning `StoppingCriteria` internally when setting the values in `RandomSearchCriteria`